gpjax/decision_making/utility_functions/probability_of_improvement.py

```diff
@@ -0,0 +1,110 @@
+    $$`\text{PI}(x) = \text{Prob}[f(x) < f(x_{\text{best}})]`$$
+    where $`x_{\text{best}}`$ is the minimizer of $`f`$ in the dataset.
+
+    The probability of improvement can be easily computed using the
+    cumulative distribution function of the standard normal distribution $`\Phi`$:
+    $$`\text{PI}(x) = \Phi\left(\frac{f(x_{\text{best}}) - \mu}{\sigma}\right)`$$
+    where $`\mu`$ and $`\sigma`$ are the mean and standard deviation of the
+    predictive distribution of the objective function at $`x`$.
+
+    References
+    ----------
+    [1] Kushner, H. J. (1964).
+    A new method of locating the maximum point of an arbitrary multipeak curve in the presence of noise.
+    Journal of Basic Engineering, 86(1), 97-106.
+
+    [2] Shahriari, B., Swersky, K., Wang, Z., Adams, R. P., & de Freitas, N. (2016).
+    Taking the human out of the loop: A review of Bayesian optimization.
+    Proceedings of the IEEE, 104(1), 148-175. doi: 10.1109/JPROC.2015.2494218
+    """
+
+    def build_utility_function(
+        self,
+        posteriors: Mapping[str, ConjugatePosterior],
+        datasets: Mapping[str, Dataset],
+        key: KeyArray,
+    ) -> SinglePointUtilityFunction:
+        """
+        Constructs the probability of improvement utility function
+        using the predictive posterior of the objective function.
+
+        Args:
+            posteriors (Mapping[str, AbstractPosterior]): Dictionary of posteriors to be
+            used to form the utility function. One of the posteriors must correspond
+            to the `OBJECTIVE` key, as we sample from the objective posterior to form
+            the utility function.
+            datasets (Mapping[str, Dataset]): Dictionary of datasets which may be used
+            to form the utility function. Keys in `datasets` should correspond to
+            keys in `posteriors`. One of the datasets must correspond
+            to the `OBJECTIVE` key.
+            key (KeyArray): JAX PRNG key used for random number generation. Since
+            the probability of improvement is computed deterministically
+            from the predictive posterior, the key is not used.
+
+        Returns:
+            SinglePointUtilityFunction: the probability of improvement utility function.
+        """
+        self.check_objective_present(posteriors, datasets)
+
+        objective_posterior = posteriors[OBJECTIVE]
+        if not isinstance(objective_posterior, ConjugatePosterior):
+            raise ValueError(
+                "Objective posterior must be a ConjugatePosterior to draw an approximate sample."
+            )
+
+        objective_dataset = datasets[OBJECTIVE]
+
+        def probability_of_improvement(x_test: Num[Array, "N D"]):
+            predictive_dist = objective_posterior.predict(x_test, objective_dataset)
+
+            # Assuming that the goal is to minimize the objective function
+            best_y = objective_dataset.y.min()
```

Comment:
It could be useful to define `best_y` as the minimum _posterior mean_ value at any of the observed points, which would also handle the case where observations could be noisy.

Reply:
Sounds good. I've updated how we compute `best_y`. It is the first time I see it being computed like that, though. If I'm not mistaken, other frameworks use the min of the dataset.

Reply:
Yeah - for instance this approach is mentioned in ["A benchmark of kriging-based infill criteria for noisy
optimization"](https://hal.science/hal-00658212/document), section 3.2.